chainermn/extensions/_multi_node_snapshot.py

```diff
@@ -0,0 +1,212 @@
+    remaining list can be ommited. Thus in that case, it can be
+    simplified more::
+
+        replica_sets = [0,]
+
+    In this case, the snapshot will be saved at rank ``0`` process and
+    at rank ``1`` process. The latter represents the replica set of
+    ``range(1, comm.size)`` . In this case autoloading at
+    initialization of snapshot extension works after the restart
+    cleanly, even though the size of the communicator differs.
+
+    Once the replica sets are defined, it can be easily extended::
+
+        replica_sets = [0,]
+        snapshot = multi_node_snapshot(comm, extensions.snapshot(),
+                                       replica_sets)
+        trainer.extend(snapshot, trigger=(1, 'epoch'))
+
+
+    More example tuples of replica set representation follows:
+
+    ===================== ===== ==============================================
+    code                  nproc actual sets
+    ===================== ===== ==============================================
+    ``[0]``               ``4`` ``[{0}, {1, 2, 3}]``
+    ``[0, 1]``            ``4`` ``[{0}, {1}, {2, 3}]``
+    ``[0, 1], [2, 3]]``   ``4`` ``[{0, 1}, {2, 3}]``
+    ``[]``                ``4`` ``[{0, 1, 2, 3}]``
+    ``[range(0, 8, 2)]``  ``8`` ``[set(range(0, 8, 2)), set(range(1, 8, 2))]``
+    ===================== ===== ==============================================
+
+    Args:
+        comm (ChainerMN communicator): communicater object
+        snapshot: Snapshot extension object obtained via
+              :meth:`~chainer.training.extensions.snapshot` .
+        replica_sets: list of replica set definition, where
+              a replica set can be defined by single integer
+              as rank number, or iterable integers.
+
+    Returns:
+        Trainer extension that wraps ``snapshot`` and properly
+        controles number of snapshots.
+
+    '''
+    return _MultiNodeSnapshot(comm, snapshot, replica_sets)
+
+
+def _parse_replica_sets(replica_sets, size):
+    sets = []
+
+    for replica_set in replica_sets:
+        if isinstance(replica_set, int):
+            assert replica_set >= 0
+            assert replica_set < size
+            sets.append({replica_set})
+        else:
+            # Must be iterable
+            for i in replica_set:
+                assert i >= 0
+                assert i < size
+            sets.append(set(replica_set))
+
+    if size > sum(len(s) for s in sets):
+        all_ranks = set(range(size))
+        all_exp = set()
+        for s in sets:
+            all_exp |= s
+        rest = all_ranks - all_exp
+        if rest:
+            sets.append(rest)
+
+    # Must guarantee: no lack allowed
+    assert size == sum(len(s) for s in sets)
+
+    # Must guarantee: no two sets must have intersection.
+    all_sum = set()
+    for s in sets:
+        all_sum |= s
+    assert size == len(all_sum)
+    return sets
+
+
+class _MultiNodeSnapshot(Extension):
+    def __init__(self, comm, snapshot, replica_sets=[]):
```

Comment:
How about using `()` as the default value of `replica_sets`?

Reply:
As actually this constructor is not directly accessed by users and `multi_node_snapshot` is the only official interface, this default argument is never used. I just removed it in following commit below.